Dynamically linking `fstat` & co

While testing `crystal i` on linux, I encountered an issue with linking `fstat` at runtime FFI (see https://github.com/crystal-lang/crystal/pull/10910#issuecomment-909380602).

`dlsym` was unable to locate the symbol `fstat` in libc.

This can be reproduced using the following code:
```cr
def load_symbol(library, name)
  handle = LibC.dlopen(nil, LibC::RTLD_LAZY | LibC::RTLD_GLOBAL)
  if handle.null?
    raise "dlopen failed for lib: #{library}: #{String.new(LibC.dlerror)}"
  end
  sym = LibC.dlsym(handle, name)
  if sym.null?
    raise "dlsym failed for lib: #{library}, name: #{name.inspect}: #{String.new(LibC.dlerror)}"
  end
  sym
end

p! load_symbol("libc", "fstat")
```

On Debian/Ubuntu, this fails with `undefined reference to `fstat'` but it works on some other Linux systems (I tried Arch and Fedora).

And indeed, `objdump` shows no `fstat` symbol in `libc.so` on Debian/Ubuntu.

`fstat` and similar methods are wrappers around internal functions like `__fxstat` (you can see that in `stat.h`). It seems that some libc instances still have symbols for those methods which makes it work. Others don't. And I haven't found an explanation for this different behaviour.

Conufisingly however, when linking a compiled program, it actually works even on Debian/Ubuntu. Even stdlib links `fstat`.

A minimal example would be this:

```cr
# crystal build --prelude=empty
lib LibC
  fun fstat(fd : Int32, buf : Void*): Int32
end

LibC.fstat(1, Pointer(Void).new(0_u64))
```

When linking with `ld` instead (`--cross-compile`, replace `cc` with `ld`, remove `-rdynamic`), the symbol can't be found either. Apparently, the linker frontend (`gcc` in my case) somehow takes care of resolving these wrapper symbols. 

So, the fact that the stdandard library bindings work for compiled programs depends on linker behaviour. In interpreted mode, we'll need to handle that differently. And it might not be a bad idea to do the same for compiled mode, because that should work the same way and we'd keep both closely aligned.

The `crystal i` branch already contains a binding to `__fxstat` as a quick workaround. This definitely works. It's not great because we're linking an internal symbol. Stability is probably not an issue, but still it's not really a documented feature.

An alternative would be to use the syscall `fstat` directly, completely ommitting `libc`. That would be possible with #10777. It might require some more work on our end, but it seems pretty straightforward actually.

--- a/src/crystal/system/unix/file.cr
+++ b/src/crystal/system/unix/file.cr
@@ -32,7 +32,18 @@ module Crystal::System::File
     if follow_symlinks
       ret = LibC.stat(path.check_no_null_byte, pointerof(stat))
     else
-      ret = LibC.lstat(path.check_no_null_byte, pointerof(stat))
+      # On some systems, the symbols `fstat` and `lstat` are not part of the GNU
+      # shared library `libc.so` and instead provided by `libc_noshared.a`.
+      # That makes them unavailable for dynamic runtime symbol lookup via `dlsym`
+      # which we use for interpreted mode.
+      # See https://github.com/crystal-lang/crystal/issues/11157#issuecomment-949640034 for details.
+      # Linking against the internal counterparts `__fxstat` and `__lxstat` directly
+      # should work in both interpreted and compiled mode.
+      {% if LibC.has_method?(:__lxstat) %}
+        ret = LibC.__lxstat(1, path.check_no_null_byte, pointerof(stat))
+      {% else %}
+        ret = LibC.lstat(path.check_no_null_byte, pointerof(stat))
+      {% end %}
     end
 
     if ret == 0

--- a/src/crystal/system/unix/file_descriptor.cr
+++ b/src/crystal/system/unix/file_descriptor.cr
@@ -64,11 +64,26 @@ module Crystal::System::FileDescriptor
   end
 
   private def system_info
-    if LibC.fstat(fd, out stat) != 0
-      raise IO::Error.from_errno("Unable to get info")
-    end
+    {% begin %}
+      # On some systems, the symbols `fstat` and `lstat` are not part of the GNU
+      # shared library `libc.so` and instead provided by `libc_noshared.a`.
+      # That makes them unavailable for dynamic runtime symbol lookup via `dlsym`
+      # which we use for interpreted mode.
+      # See https://github.com/crystal-lang/crystal/issues/11157#issuecomment-949640034 for details.
+      # Linking against the internal counterparts `__fxstat` and `__lxstat` directly
+      # should work in both interpreted and compiled mode.
+      {% if LibC.has_method?(:__fxstat) %}
+        ret = LibC.__fxstat(1, fd, out stat)
+      {% else %}
+        ret = LibC.fstat(fd, out stat)
+      {% end %}
 
-    FileInfo.new(stat)
+      if ret != 0
+        raise IO::Error.from_errno("Unable to get info")
+      end
+
+      FileInfo.new(stat)
+    {% end %}
   end
 
   private def system_seek(offset, whence : IO::Seek) : Nil

--- a/src/lib_c/x86_64-linux-gnu/c/sys/stat.cr
+++ b/src/lib_c/x86_64-linux-gnu/c/sys/stat.cr
@@ -47,6 +47,8 @@ lib LibC
   fun chmod(file : Char*, mode : ModeT) : Int
   fun fstat(fd : Int, buf : Stat*) : Int
   fun lstat(file : Char*, buf : Stat*) : Int
+  fun __fxstat(ver : Int, fd : Int, buf : Stat*) : Int
+  fun __lxstat(ver : Int, file : Char*, buf : Stat*) : Int
   fun mkdir(path : Char*, mode : ModeT) : Int
   fun mkfifo(path : Char*, mode : ModeT) : Int
   fun mknod(path : Char*, mode : ModeT, dev : DevT) : Int

